controllers/imagerepository_controller_test.go

```diff
@@ -1226,6 +1226,73 @@ var _ = Describe("Image repository controller", func() {
 		})
 	})
 
+	Context("Skip quay repository deletion", func() {
+		var resourceKey = types.NamespacedName{Name: defaultImageRepositoryName + "-skip-delete", Namespace: defaultNamespace}
+
+		BeforeEach(func() {
+			quay.ResetTestQuayClient()
+			deleteImageRepository(resourceKey)
+		})
+
+		It("should prepare environment", func() {
+			createServiceAccount(defaultNamespace, buildPipelineServiceAccountName)
+		})
+
+		It("don't remove repository if explicitly requested using annotation", func() {
+			customImageName := defaultNamespace + "/" + "my-image-should-not-be-removed"
+			expectedImageName = customImageName
+			expectedRobotAccountPrefix = strings.ReplaceAll(strings.ReplaceAll(expectedImageName, "-", "_"), "/", "_")
+
+			isCreateRepositoryInvoked := false
+			quay.CreateRepositoryFunc = func(repository quay.RepositoryRequest) (*quay.Repository, error) {
+				defer GinkgoRecover()
+				isCreateRepositoryInvoked = true
+				Expect(repository.Repository).To(Equal(expectedImageName))
+				Expect(repository.Namespace).To(Equal(quay.TestQuayOrg))
+				return &quay.Repository{Name: expectedImageName}, nil
+			}
+
+			createImageRepository(imageRepositoryConfig{
+				ImageName:   customImageName,
+				ResourceKey: &resourceKey,
+				Annotations: map[string]string{SkipRepositoryDeletionAnnotationName: "true"},
+			})
+			Eventually(func() bool { return isCreateRepositoryInvoked }, timeout, interval).Should(BeTrue())
+			waitImageRepositoryFinalizerOnImageRepository(resourceKey)
+
+			imageRepository := getImageRepository(resourceKey)
+			Expect(imageRepository.Spec.Image.Name).To(Equal(expectedImageName))
+
+			isDeleteRobotAccountInvoked := false
+			quay.DeleteRobotAccountFunc = func(organization, robotAccountName string) (bool, error) {
+				defer GinkgoRecover()
+				isDeleteRobotAccountInvoked = true
+				Expect(organization).To(Equal(quay.TestQuayOrg))
+				Expect(strings.HasPrefix(robotAccountName, expectedRobotAccountPrefix)).To(BeTrue())
+				return true, nil
+			}
+
+			isDeleteRepositoryInvoked := false
+			quay.DeleteRepositoryFunc = func(organization, imageRepository string) (bool, error) {
+				defer GinkgoRecover()
+				isDeleteRepositoryInvoked = true
+				Expect(organization).To(Equal(quay.TestQuayOrg))
+				Expect(imageRepository).To(Equal(customImageName))
+				return true, nil
+			}
+
+			deleteImageRepository(resourceKey)
+			// should not delete repository, because of the annotation
+			waitImageRepositoryGone(resourceKey)
+			Eventually(func() bool { return isDeleteRobotAccountInvoked }, timeout, interval).Should(BeTrue())
+			Eventually(func() bool { return isDeleteRepositoryInvoked }, timeout, interval).Should(BeFalse())
```

Comment:
I'm not sure if this really tests anything. `isDeleteRepositoryInvoked` starts out as `false`, so I think this `Eventually()` will just succeed immediately, even though the repository could still get deleted later

Reply:
the call to `deleteImageRepository(resourceKey)` will ensure that the image repository resources was deleted. It means that the call to remove the finalizer was made, and this happens after the function for deleting the external repository is called.

I removed the use of `Eventually` and cleaned the mock function for `DeleteRepositoryFunc`.   I hope it's less confusing now.

Reply:
Ack, makes sense